File: hecho con python/app/services/estudiante_service.py
from typing import List
from fastapi import HTTPException, status
from app.models.estudiante import Estudiante
from app.dtos.estudiante_dto import EstudianteCreateDTO, EstudianteUpdateDTO, EstudianteResponseDTO
from app.repositories.estudiante_repository import EstudianteRepository

class EstudianteService:
    def __init__(self, repository: EstudianteRepository):
        self.repository = repository

    def obtener_todos_los_estudiantes(self) -> List[EstudianteResponseDTO]:
        estudiantes = self.repository.get_all()
        return [EstudianteResponseDTO.model_validate(e) for e in estudiantes]

    def obtener_estudiante_por_id(self, estudiante_id: int) -> EstudianteResponseDTO:
        estudiante = self.repository.get_by_id(estudiante_id)
        if not estudiante:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Estudiante con ID {estudiante_id} no encontrado"
            )
        return EstudianteResponseDTO.model_validate(estudiante)

    def obtener_estudiante_por_correo(self, correo: str) -> EstudianteResponseDTO:
        estudiante = self.repository.get_by_correo(correo)
        if not estudiante:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Estudiante con correo '{correo}' no encontrado"
            )
        return EstudianteResponseDTO.model_validate(estudiante)

    def crear_estudiante(self, dto: EstudianteCreateDTO) -> EstudianteResponseDTO:
        # Validación de regla de negocio: Correo único
        if self.repository.get_by_correo(dto.correo):
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"Ya existe un estudiante registrado con el correo '{dto.correo}'"
            )
        
        nuevo_estudiante = Estudiante(
            nombre=dto.nombre,
            telefono=dto.telefono,
            correo=dto.correo
        )
        guardado = self.repository.create(nuevo_estudiante)
        return EstudianteResponseDTO.model_validate(guardado)

    def actualizar_estudiante(self, estudiante_id: int, dto: EstudianteUpdateDTO) -> EstudianteResponseDTO:
        estudiante = self.repository.get_by_id(estudiante_id)
        if not estudiante:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Estudiante con ID {estudiante_id} no encontrado para actualizar"
            )
        
        if dto.correo and dto.correo != estudiante.correo:
            if self.repository.get_by_correo(dto.correo):
                raise HTTPException(
                    status_code=status.HTTP_409_CONFLICT,
                    detail=f"El correo '{dto.correo}' ya está en uso por otro estudiante"
                )
            estudiante.correo = dto.correo
            
        if dto.nombre is not None:
            estudiante.nombre = dto.nombre
        if dto.telefono is not None:
            estudiante.telefono = dto.telefono

        actualizado = self.repository.update(estudiante)
        return EstudianteResponseDTO.model_validate(actualizado)

File: hecho con java/src/main/java/com/umanizales/notasya/services/EstudianteService.java
package com.umanizales.notasya.services;

import com.umanizales.notasya.dtos.EstudianteDTO;
import com.umanizales.notasya.exceptions.ConflictException;
import com.umanizales.notasya.exceptions.ResourceNotFoundException;
import com.umanizales.notasya.models.Estudiante;
import com.umanizales.notasya.repositories.EstudianteRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class EstudianteService {

    private final EstudianteRepository repository;

    public EstudianteService(EstudianteRepository repository) {
        this.repository = repository;
    }

    @Transactional(readOnly = true)
    public List<EstudianteDTO.Response> obtenerTodosLosEstudiantes() {
        return repository.findAll().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public EstudianteDTO.Response obtenerEstudiantePorId(Long id) {
        Estudiante e = repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Estudiante con ID " + id + " no encontrado"));
        return mapToResponse(e);
    }

    @Transactional(readOnly = true)
    public EstudianteDTO.Response obtenerEstudiantePorCorreo(String correo) {
        Estudiante e = repository.findByCorreo(correo)
                .orElseThrow(() -> new ResourceNotFoundException("Estudiante con correo '" + correo + "' no encontrado"));
        return mapToResponse(e);
    }

    @Transactional
    public EstudianteDTO.Response crearEstudiante(EstudianteDTO.Create dto) {
        if (repository.existsByCorreo(dto.getCorreo())) {
            throw new ConflictException("Ya existe un estudiante registrado con el correo: " + dto.getCorreo());
        }

        Estudiante e = Estudiante.builder()
                .nombre(dto.getNombre())
                .telefono(dto.getTelefono())
                .correo(dto.getCorreo())
                .build();

        Estudiante guardado = repository.save(e);
        return mapToResponse(guardado);
    }

    @Transactional
    public EstudianteDTO.Response actualizarEstudiante(Long id, EstudianteDTO.Update dto) {
        Estudiante e = repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Estudiante con ID " + id + " no encontrado"));

        if (dto.getCorreo() != null && !dto.getCorreo().equals(e.getCorreo())) {
            if (repository.existsByCorreo(dto.getCorreo())) {
                throw new ConflictException("El correo '" + dto.getCorreo() + "' ya está en uso");
            }
            e.setCorreo(dto.getCorreo());
        }

        if (dto.getNombre() != null) e.setNombre(dto.getNombre());
        if (dto.getTelefono() != null) e.setTelefono(dto.getTelefono());

        Estudiante actualizado = repository.save(e);
        return mapToResponse(actualizado);
    }

    private EstudianteDTO.Response mapToResponse(Estudiante e) {
        return EstudianteDTO.Response.builder()
                .id(e.getId())
                .nombre(e.getNombre())
                .telefono(e.getTelefono())
                .correo(e.getCorreo())
                .build();
    }
}
